urld829b80034_wjch111_PTSP2013Wang_tgz-pJ8-controllers_momcts_momctsCore_TestPlannerJ8/src/framework/core/Game.java

```diff
@@ -0,0 +1,726 @@
+
+        //Number of waypoints to be collected.
+        m_waypointsLeft = m_numWaypoints;
+
+        //number of available fuel tanks
+        m_fuelTanksLeft = m_numFuelTanks;
+
+        //This is the number of steps allowed until reaching the next waypoint.
+        m_stepsLeft = PTSPConstants.getStepsPerWaypoints(m_numWaypoints);
+
+        //Create the ship of the game and add it to the game objects
+        m_ship = new Ship(this, getMap().getStartingPoint());
+        m_gameObjects.add(m_ship);
+    }
+
+    /**
+     * Updates the game logic: all entities in the game and game state.
+     * @param a_shipMove Move to execute.
+     */
+    public void tick(int a_shipMove)
+    {
+        //Set the next move of the player
+        m_ship.setNextMove(a_shipMove);
+
+        //Update all entities of the game, including player's ship.
+        synchronized (Game.class) {
+            for (GameObject ob : m_gameObjects) {
+                ob.update();
+            }
+        }
+
+        //One step left to the end.
+        if(m_started)
+        {
+            m_stepsLeft--;
+            m_totalTime++;
+        }
+
+        //Check for end of the game.
+        if(checkGameEnded())
+            m_gameEnded = true;
+
+    }
+
+    /**
+     * Returns true if the game has ended because the time has run out, all waypoints have been collected or
+     * the ship is broken.
+     * @return  true if the game has ended
+     */
+    private boolean checkGameEnded()
+    {
+        return (m_waypointsLeft == 0 || m_stepsLeft <= 0 || m_ship.getDamage() >= PTSPConstants.MAX_DAMAGE);
+    }
+
+
+    /**
+     * Updates the game when a waypoint is collected.
+     */
+    public void waypointCollected()
+    {
+        m_stepsLeft = PTSPConstants.getStepsPerWaypoints(m_numWaypoints);
+        m_waypointsLeft--;
+    }
+
+    /**
+     * Updates the game when a fuel tank is collected.
+     */
+    public void fuelTankCollected()
+    {
+        m_ship.addFuel(PTSPConstants.FUEL_TANK_BOOST);
+        m_fuelTanksLeft--;
+    }
+
+    /**
+     * Adds a waypoint to the list of collected
+     * @param a_index Index in the array list of waypoints of the waypoint collected.
+     */
+    public void addCollected(int a_index)
+    {
+        m_visitOrder.add(a_index);
+    }
+
+    /**
+     * Saves a route to file.
+     */
+    public void saveRoute()
+    {
+        String routeFile = getSaveFilename();
+        try {
+            PrintWriter out = new PrintWriter(new FileWriter(routeFile));
+            out.println(m_ship.getActionList().size());
+            for (int i = 0; i < m_ship.getActionList().size(); i++) {
+                out.println(m_ship.getActionList().get(i));
+            }
+            out.close();
+            System.out.println("Route saved to: " + routeFile);
+        } catch (Exception e) {
+            System.out.println("Exception saving route: " + routeFile);
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 388:398, 409:419.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.